core/src/main/kotlin/org/eln2/math/FunctionTable.kt

```diff
@@ -3,38 +3,51 @@ package org.eln2.math
 /**
  * Function Table
  *
- * Defines a function based on the values in a table.
+ * Defines a function based on the values in a table. The domain of the input array starts from 0 and goes to "xMax"
+ * @param point Array of double representing the reference or "y" values.
+ * @param xMax The x value that corresponds to the last y value in the point array.
  */
-open class FunctionTable(var point: DoubleArray, var xMax: Double) : IFunction {
-	var xMaxInv: Double
-	var xDelta: Double
+open class FunctionTable(var point: DoubleArray, var xMax: Double, private val mode : ExtrapolationMode) : IFunction {
+
+	/**
+	 * Selects which method of extrapolation to use when trying to get a value outside of the given range.
+	 */
+	enum class ExtrapolationMode {
+		ClosestValue, Linear
+	}
 
 	/**
 	 * getValue: Given a value, get the corresponding value, linearly fit between two closest points on either side.
+	 * Note that this is an interpolation. Extrapolation will return null.
 	 * @param x: Given value
-	 * @return y: Value for that x, linearly fit between the two closest points on either side
+	 * @return y: Value for that x, linearly fit between the two closest points on either side if valid; otherwise
+	 * 	extrapolate using the method specified when creating this FunctionTable.
 	 */
 	override fun getValue(x: Double): Double {
-		var lx = x
-		lx *= xMaxInv
-		if (lx < 0f) return point[0] + (point[1] - point[0]) * (point.size - 1) * lx
-		if (lx >= 1.0f) return point[point.size - 1] + (point[point.size - 1] - point[point.size - 2]) * (point.size - 1) * (lx - 1.0)
-		lx *= point.size - 1.toDouble()
-		val idx = lx.toInt()
-		lx -= idx.toDouble()
-		return point[idx + 1] * lx + point[idx] * (1.0f - lx)
-	}
-
-	/*public double getValue(double x) {
-		double a = getValueLin(-xDelta);
-		double b = getValueLin(xDelta);
-		double firFactorA = 0.5, firFactorB = (1 - firFactorA) / 2;
-		return getValueLin(x) * firFactorA + getValueLin(x - xDelta) * firFactorB + getValueLin(x + xDelta) * firFactorB;
+		val lx = x.div(xMax)
+		return when {
+			lx < 0.0 -> { // Extrapolation to the left.
+				when (mode) {
+					ExtrapolationMode.ClosestValue -> point[0]
+					ExtrapolationMode.Linear -> point[0] + (point[1] - point[0]) * (point.size - 1) * lx
+				}
+			}
+			lx > 1.0 -> { // Extrapolation to the right.
+				when (mode) {
+					ExtrapolationMode.ClosestValue -> point.last()
+					ExtrapolationMode.Linear -> point.last() + (point.last() - point[point.size - 2]) * point.last() * (lx - 1.0)
+				}
+			}
+			else -> { // Interpolation
+				val sx = lx.times(point.size - 1)
```

Comment:
just curious: why the non-operator `times` here?

Reply:
My guess is that when IDEA converted this file from Java to Kotlin that it thought to keep this? That's my best guess, but it's not necessary to keep this way.